src/responder/common/cache_req/cache_req.c

```diff
@@ -0,0 +1,558 @@
+/*
+    Authors:
+        Pavel Březina <pbrezina@redhat.com>
+
+    Copyright (C) 2016 Red Hat
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <ldb.h>
+#include <talloc.h>
+#include <tevent.h>
+
+#include "util/util.h"
+#include "responder/common/cache_req/cache_req_private.h"
+#include "responder/common/cache_req/cache_req_plugin.h"
+
+static struct cache_req_plugin *
+cache_req_get_plugin(enum cache_req_type type)
+{
+    static struct cache_req_plugin *plugins[CACHE_REQ_SENTINEL] = {
+
+    };
+
+    if (type >= CACHE_REQ_SENTINEL) {
+        DEBUG(SSSDBG_FATAL_FAILURE, "Bug: invalid plugin type!");
+        return NULL;
+    }
+
+    return plugins[type];
+}
+
+static void cache_req_set_plugin(struct cache_req *cr,
+                                 enum cache_req_type type)
+{
+    struct cache_req_plugin *plugin;
+
+    plugin = cache_req_get_plugin(type);
+    if (plugin == NULL) {
+        DEBUG(SSSDBG_FATAL_FAILURE, "Bug: unset plugin!");
+        return;
+    }
+
+    cr->reqname = plugin->name;
+    cr->dp_type = plugin->dp_type;
+    cr->plugin = plugin;
+
+    CACHE_REQ_DEBUG(SSSDBG_TRACE_INTERNAL, cr, "Setting \"%s\" plugin\n",
+                    plugin->name);
+
+    return;
+}
+
+static struct cache_req *
+cache_req_create(TALLOC_CTX *mem_ctx,
+                 struct resp_ctx *rctx,
+                 struct cache_req_data *data,
+                 struct sss_nc_ctx *ncache,
+                 int midpoint)
+{
+    struct cache_req *cr;
+
+    cr = talloc_zero(mem_ctx, struct cache_req);
+    if (cr == NULL) {
+        return NULL;
+    }
+
+    cr->rctx = rctx;
+    cr->data = data;
+    cr->ncache = ncache;
+    cr->midpoint = midpoint;
+    cr->req_start = time(NULL);
```

Comment:
Initial timestamp is basically logged in `cache_req_send()` with "New request" debug message. Isn't this enough? If midpoint refresh is triggered it is also logged.